online-shop/src/ReduxStore/ShoppingCartSection/reducers.tsx

```diff
@@ -0,0 +1,170 @@
+                productsInShoppingCart: state.productsInShoppingCart,
+                uniqueProductsInShoppingCart: state.uniqueProductsInShoppingCart,
+                checkoutActionStatus: 0,
+                numberOfProductsInShoppingCart: calculateNumberOfItemsInShoppingCart(state.productsInShoppingCart),
+                totalPriceForShoppingCart: calculateTotalPriceForShoppingCart(state.productsInShoppingCart)
+            };
+        }
+
+        default:
+            return state;
+    }
+}
+
+let calculateNumberOfSameItem = (arrayToCalculateFrom: IProduct[], productId: number): number => {
+    let numberOfSameProduct = 0;
+
+    arrayToCalculateFrom.forEach(
+        (product) => {
+            if (product.id === productId) {
+                numberOfSameProduct++;
+            }
+        });
+
+    return numberOfSameProduct;
+}
+
+let calculateNumberOfItemsInShoppingCart = (arrayToCalculateFrom: IProduct[]): number => {
+    return arrayToCalculateFrom.length;
+}
+
+let calculateTotalPriceForShoppingCart = (arrayToCalculateFrom: IProduct[]): number => {
+    let calculatedPrice = 0;
+
+    if (arrayToCalculateFrom.length) {
+        arrayToCalculateFrom.forEach(
+            (currentProduct) => {
+                calculatedPrice += currentProduct.price;
+            }
+        );
+    }
+
+    return calculatedPrice;
+}
+
+let removeDuplicatesFromMainArray = (state: SCTypes.ShoppingCartState = initialState, arrayToParse: IProduct[]): IProduct[] => {
+    //pentru fiecare element din state
+    arrayToParse.forEach((product) => {
+        let flagProductAlreadyExists: boolean = true;
+
+        //verificam daca este in unique, in cazul in care unique are macar 1 element
+        if (state.uniqueProductsInShoppingCart.length >= 1) {
+            state.uniqueProductsInShoppingCart.forEach((uniqueProduct) => {
+                if (product.id === uniqueProduct.id) {
+                    flagProductAlreadyExists = false;
+                }
+            });
+        }
+
+        if (flagProductAlreadyExists) {
+            state.uniqueProductsInShoppingCart = [...state.uniqueProductsInShoppingCart, product];
+        }
+    });
+
+    return state.uniqueProductsInShoppingCart;
+}
+
+let decreaseProductQuantityFromShoppingCart = (arrayToDecreaseQuantityFrom: IProduct[], productId: number): IProduct[] => {
+    let shoppingArrayIndex = arrayToDecreaseQuantityFrom.findIndex(i => i.id === productId);
+    let temporaryArray: IProduct[] = arrayToDecreaseQuantityFrom;
+
+    if (shoppingArrayIndex > -1) {
+        //am pus asta aici si merge acum
+        temporaryArray.splice(shoppingArrayIndex, 1);
+    }
+
+    return temporaryArray;
+}
+
+let removeItemFromUniqueArray = (state: SCTypes.ShoppingCartState = initialState, productID: number): IProduct[] => {
+    let productIndexInUniqueArray = state.uniqueProductsInShoppingCart.findIndex(i => i.id === productID);
+    let temporaryArray: IProduct[] = state.uniqueProductsInShoppingCart;
```

Comment:
You are actually not copying the array here in a new array. Both variables (`temporaryArray` and `state.uniqueProductsInShoppingCart`) point to the same array - so you spliced the array from the state - which is bad. You should not change the old state in the reducer.